crates/test-utils/src/tx_indexer.rs

```diff
@@ -0,0 +1,350 @@
+    bitcoin::{
+        build_test_deposit_request_script, build_test_deposit_script, create_test_deposit_tx,
+        generate_withdrawal_fulfillment_data, test_taproot_addr,
+    },
+    l2::gen_params,
+};
+
+// TEST FUNCTIONS
+
+/// Runs a test deposit transaction with the given parameters and returns the indexer's output.
+/// The caller can then perform further tests on the output.
+pub fn test_index_multiple_deposits_with_visitor<V>(
+    visitor: impl Fn() -> V,
+    ops_extractor: fn(&Indexed<V::Output>) -> Vec<ProtocolOperation>,
+) -> Vec<Indexed<V::Output>>
+where
+    V: TxVisitor,
+{
+    let params = gen_params();
+    let (filter_config, keypair) = create_tx_filter_config(&params);
+    let deposit_config = filter_config.deposit_config.clone();
+    let idx1 = 0xdeafbeef;
+    let idx2 = 0x1badb007;
+    let dest_addr1 = vec![3u8; 20];
+    let dest_addr2 = vec![4u8; 20];
+    let tapnodehash = [0u8; 32];
+
+    let deposit_script1 =
+        build_test_deposit_script(&deposit_config, idx1, dest_addr1.clone(), &tapnodehash);
+    let deposit_script2 =
+        build_test_deposit_script(&deposit_config, idx2, dest_addr2.clone(), &tapnodehash);
+
+    let tx1 = create_test_deposit_tx(
+        Amount::from_sat(deposit_config.deposit_amount),
+        &deposit_config.address.address().script_pubkey(),
+        &deposit_script1,
+        &keypair,
+        &tapnodehash,
+    );
+    let tx2 = create_test_deposit_tx(
+        Amount::from_sat(deposit_config.deposit_amount),
+        &deposit_config.address.address().script_pubkey(),
+        &deposit_script2,
+        &keypair,
+        &tapnodehash,
+    );
+
+    let block = create_test_block(vec![tx1, tx2]);
+
+    let tx_entries = index_block(&block, visitor, &filter_config);
+
+    assert_eq!(
+        tx_entries.len(),
+        2,
+        "test: should find two relevant transactions"
+    );
+
+    for (i, info) in tx_entries.iter().flat_map(ops_extractor).enumerate() {
+        if let ProtocolOperation::Deposit(deposit_info) = info {
+            assert_eq!(
+                deposit_info.address,
+                if i == 0 {
+                    dest_addr1.clone()
+                } else {
+                    dest_addr2.clone()
+                },
+                "test: dest should match for transaction {i}",
+            );
+            assert_eq!(
+                deposit_info.deposit_idx,
+                [idx1, idx2][i],
+                "test: deposit idx should match"
+            );
+            assert_eq!(
+                deposit_info.amt,
+                BitcoinAmount::from_sat(deposit_config.deposit_amount),
+                "test: deposit amount should match for transaction {i}",
+            );
+        } else {
+            panic!("test: expected DepositInfo for transaction {i}");
+        }
+    }
+
+    tx_entries
+}
+
+pub fn test_index_no_deposit_with_visitor<V>(
+    visitor: impl Fn() -> V,
+    _: fn(&Indexed<V::Output>) -> Vec<ProtocolOperation>,
```

Comment:
Probably should be `impl Fn` so that it's more general, we don't care about it being just a fn ptr.  Same in the other test fns here.